### ISSUE TITLE
[ProactiveThreadpool Bug] shutdown hangs

### ISSUE THREAD

**Issue by twhuang-uiuc** [cpp-taskflow/cpp-taskflow#42]
The proactive thread pool hangs during dynamic tasking - task adds another tasks to the threadpool. For example, the following code can cause proactive threadpool to hang when calling shutdown:

```cpp
std::function<void(int)> insert;
insert = [&threadpool, &insert, &sum] (int i) {
  if(i > 0) {
    threadpool.silent_async([i=i-1, &insert] () {
      insert(i);
    }); 
  }   
  else {
    if(auto s = ++sum; s == num_threads) {
      promise.set_value();
    }   
  }   
};

// emplace dynamic tasks
for(size_t i=0; i<num_threads; i++){
  insert(num_tasks / num_threads);
}
  
// this will hang
threadpool.shutdown();
```

**Comment by gguo4** [comment]
If you enable dynamic tasking, all parents tasks can be completed and freed from the working loop, but child tasks which are created after the master thread cleared shared class members don't have the environment to finish their execution. 

I have fixed it with an update. Upon calling shutdown, the proactive thread pool won't release its workers until all threads are idle.

**Comment by twhuang-uiuc** [comment]
This is a perfect solution. Merged to the dev branch (#43). Thank you!
